bindings/rabbitmq/rabbitmq.go

```diff
@@ -230,66 +291,153 @@ func (r *RabbitMQ) declareQueue() (amqp.Queue, error) {
 	}
 
-	msgs, err := r.channel.Consume(
-		r.queue.Name,
-		"",
-		false,
-		false,
-		false,
-		false,
-		nil,
-	)
-	if err != nil {
-		return err
-	}
-
 	readCtx, cancel := context.WithCancel(ctx)
-
 	r.wg.Add(2)
 	go func() {
-		defer r.wg.Done()
-		defer cancel()
 		select {
 		case <-r.closeCh:
+			// nop
 		case <-readCtx.Done():
+			// nop
 		}
+		r.wg.Done()
+		cancel()
 	}()
-
 	go func() {
+		// unless closed, keep trying to read and handle messages forever
 		defer r.wg.Done()
-		var err error
 		for {
-			select {
-			case <-readCtx.Done():
+			var (
+				msgs              <-chan amqp.Delivery
+				err               error
+				declaredQueueName string
+				ch                *amqp.Channel
+			)
+			r.channelMutex.RLock()
+			declaredQueueName = r.queue.Name
+			ch = r.channel
+			r.channelMutex.RUnlock()
+			if ch == nil {
+				goto Retry
+			}
+			msgs, err = ch.Consume(
+				declaredQueueName,
+				"",
+				false,
+				false,
+				false,
+				false,
+				nil,
+			)
+			if err != nil {
+				r.logger.Errorf("Error consuming messages from queue [%s]: %v", r.queue.Name, err)
+				goto Retry
+			}
+			r.handleMessage(readCtx, handler, msgs, ch)
+
+		Retry:
+			if r.closed.Load() || readCtx.Err() != nil {
+				r.logger.Info("Input binding closed, stop fetching message")
 				return
-			case d := <-msgs:
-				_, err = handler(readCtx, &bindings.ReadResponse{
-					Data: d.Body,
-				})
-				if err != nil {
-					r.channel.Nack(d.DeliveryTag, false, true)
-				} else {
-					r.channel.Ack(d.DeliveryTag, false)
-				}
 			}
+
+			time.Sleep(r.metadata.reconnectWait)
 		}
 	}()
-
 	return nil
 }
 
+// handleMessage handles incoming messages from RabbitMQ
+func (r *RabbitMQ) handleMessage(ctx context.Context, handler bindings.Handler, msgCh <-chan amqp.Delivery, ch *amqp.Channel) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case d, ok := <-msgCh:
+			if !ok {
+				r.logger.Info("Input binding channel closed")
+				return
+			}
+			_, err := handler(ctx, &bindings.ReadResponse{
+				Data: d.Body,
+			})
+			if err != nil {
+				ch.Nack(d.DeliveryTag, false, true)
+			} else {
+				ch.Ack(d.DeliveryTag, false)
+			}
+		}
+	}
+}
+
 func (r *RabbitMQ) Close() error {
 	if r.closed.CompareAndSwap(false, true) {
 		close(r.closeCh)
 	}
 	defer r.wg.Wait()
-	return r.channel.Close()
+	r.channelMutex.Lock()
+	defer r.channelMutex.Unlock()
+	return r.reset()
+}
+
+func (r *RabbitMQ) connect() error {
+	if r.closed.Load() {
+		// Do not reconnect on stopped service.
+		return errClosed
+	}
+
+	conn, ch, err := dial(r.metadata.Host)
+	if err != nil {
+		return err
+	}
+
+	ch.Qos(r.metadata.PrefetchCount, 0, true)
+	q, err := r.declareQueue(ch)
+	if err != nil {
+		return err
+	}
+
+	r.notifyRabbitChannelClose = make(chan *amqp.Error, 1)
+	ch.NotifyClose(r.notifyRabbitChannelClose)
+
+	r.channelMutex.Lock()
+	// try to close the old channel and connection, ignore the error
+	_ = r.reset() //nolint:errcheck
```

Comment:
Do we not want to at least log this error message?

Reply:
The error message is logged inside the `reset` function. So here do not need logs.